extensions-core/protobuf-extensions/src/test/java/org/apache/druid/data/input/protobuf/ProtobufReaderTest.java

```diff
@@ -0,0 +1,216 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input.protobuf;
+
+import com.google.common.collect.Lists;
+import com.google.protobuf.Timestamp;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.InputRowSchema;
+import org.apache.druid.data.input.impl.DimensionsSpec;
+import org.apache.druid.data.input.impl.StringDimensionSchema;
+import org.apache.druid.data.input.impl.TimestampSpec;
+import org.apache.druid.java.util.common.parsers.JSONPathFieldSpec;
+import org.apache.druid.java.util.common.parsers.JSONPathFieldType;
+import org.apache.druid.java.util.common.parsers.JSONPathSpec;
+import org.joda.time.DateTime;
+import org.joda.time.chrono.ISOChronology;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+public class ProtobufReaderTest
+{
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  private InputRowSchema inputRowSchema;
+  private InputRowSchema inputRowSchemaWithComplexTimestamp;
+  private JSONPathSpec flattenSpec;
+  private FileBasedProtobufBytesDecoder decoder;
+
+  @Before
+  public void setUp()
+  {
+    TimestampSpec timestampSpec = new TimestampSpec("timestamp", "iso", null);
+    DimensionsSpec dimensionsSpec = new DimensionsSpec(Lists.newArrayList(
+        new StringDimensionSchema("event"),
+        new StringDimensionSchema("id"),
+        new StringDimensionSchema("someOtherId"),
+        new StringDimensionSchema("isValid")
+    ), null, null);
+    flattenSpec = new JSONPathSpec(
+        true,
+        Lists.newArrayList(
+            new JSONPathFieldSpec(JSONPathFieldType.ROOT, "eventType", "eventType"),
+            new JSONPathFieldSpec(JSONPathFieldType.PATH, "foobar", "$.foo.bar"),
+            new JSONPathFieldSpec(JSONPathFieldType.PATH, "bar0", "$.bar[0].bar")
+        )
+    );
+
+    inputRowSchema = new InputRowSchema(timestampSpec, dimensionsSpec, null);
+    inputRowSchemaWithComplexTimestamp = new InputRowSchema(
+        new TimestampSpec("otherTimestamp", "iso", null),
+        dimensionsSpec,
+        null
+    );
+    decoder = new FileBasedProtobufBytesDecoder("prototest.desc", "ProtoTestEvent");
+  }
+
+  @Test
+  public void testParseNestedData() throws Exception
+  {
+    ProtobufReader reader = new ProtobufReader(inputRowSchema, null, decoder, flattenSpec);
+
+    //create binary of proto test event
+    DateTime dateTime = new DateTime(2012, 7, 12, 9, 30, ISOChronology.getInstanceUTC());
+    ProtoTestEventWrapper.ProtoTestEvent event = ProtoTestEventWrapper.ProtoTestEvent.newBuilder()
```

Comment:
<nit> can probably create a template event builder and a template assertion block to avoid so much duplication

Reply:
that's a good suggestion. I have refactored the test code so that even building and assertion are now reused.